pycobertura/cobertura.py

```diff
@@ -47,7 +49,15 @@ def __init__(self, report, filesystem=None):
         files referenced in the report. Please check the pycobertura.filesystem
         module in order to discover more about filesystems.
         """
-        self.xml = ET.parse(report).getroot()
+        try:
+            self.xml = ET.fromstring(report).getroot()
+        except lxml.etree.XMLSyntaxError:
+            # it's not a xml-string, so it should be a path.
+            # is it existed?
+            if not os.path.isfile(report):
+                raise FileNotFoundError("file not existed: {}".format(report))
+            # it is an existed file.
+            self.xml = ET.parse(report).getroot()
```

Comment:
The method `__init__()` is getting a little crowded... Can we break out the logic into methods such as `_load_from_string` and `_load_from_path`?

```suggestion
        try:
            self.xml = self._load_from_string(report)
        except lxml.etree.XMLSyntaxError:
            self.xml = self._load_from_path(report)
```

... and then abstract away the logic in the respective methods. What do you think?

Reply:
Approve. I prefer making them a closure because they will only be used in this method:

```python
def _load_from_string() ...
def _load_from_path() ...

try:
    self.xml = _load_from_string(report)
except lxml.etree.XMLSyntaxError:
    self.xml = _load_from_path(report)
```

Both are ok I think.

Reply:
Hm. I was trying to visually declutter `__init__()`, so making them closures would defeat my initial intent.

In Python, I tend to use closures to prevent me from writing small classes, to carry state when I'm only using functions. But here we're already in the context of a full-fledged class so would adding closures add much value? If the purpose is because they aren't used anywhere else, I actually like to create functions/methods that aren't necessarily shared. I think they bring additional readability to the code, thanks to their function's name. I'd rather reuse the existing mechanics of the class instead of introducing a pattern of closures in this code.

What do you think?